src/accel/AlongStepFactory.hh

```diff
@@ -84,6 +84,12 @@ class AlongStepFactoryInterface
 
   public:
     virtual ~AlongStepFactoryInterface() = default;
+    AlongStepFactoryInterface() = default;
+    AlongStepFactoryInterface(AlongStepFactoryInterface const&) = delete;
+    AlongStepFactoryInterface& operator=(AlongStepFactoryInterface const&)
+        = delete;
+    AlongStepFactoryInterface(AlongStepFactoryInterface&&) = delete;
+    AlongStepFactoryInterface& operator=(AlongStepFactoryInterface&&) = delete;
```

Comment:
yeah for abstract base class, I'm not convinced it is worth it to add all this verbosity in the declaration. You can still run unto issues, e.g. this code:

```c++
class B {
public:
    virtual char m() = 0;
    B() = default;
    // B(const B&) = delete;
    // B& operator=(const B&) = delete;
};

class D : public B {
public:
    char m() override { return 'D'; }
    // ...
};

class E : public B {
public:
    char m() override { return 'E'; }
    // ...
};


int main() {
    D d;
    E e;
    B* b = &d;
    std::cout << b->m() << std::endl; // prints D
    *b = e; // should not use value semantics but allowed because we have a copy assignment, b = &e
    std::cout << b->m() << std::endl;  // prints D
    return 0;
}
```

Reply:
Ah true. I wonder in what use cases polymorphic assignment *does* work then? I favor the less verbose approach for ABCs.

Reply:
>  I wonder in what use cases polymorphic assignment does work then? 

With a `virtual operator=`? (or which a generic `operator=` with internal dispatch (i.e. dynamic_cast, i.e. likely a bad idea but still).

Reply:
> for pure abstract base classes (where there's no data to be copied or moved)

The 'abstracted' does not guarantee that it will stay pure.  i.e. the standard allow abstract class with data members.

Reply:
As a rule the "Interface" classes in Celeritas are pure virtual with no data.

Reply:
Fair enough.  Still the [example](https://github.com/celeritas-project/celeritas/pull/834#discussion_r1245382286) applies to Celeritas and we know we never want:
```
    *b = e; // should not use value semantics but allowed because we have a copy assignment, b = &e
```
and thus we ought to try to make it not compilable, isn't it?

Reply:
I hope a code review would catch assignment of an interface class... it just seems like a very unusual mistake and not worth us trying to guard against.

Reply:
Maybe for the user-facing interface in `accel` it's worth adding it?

Reply:
> I hope a code review would catch assignment of an interface class... it just seems like a very unusual mistake and not worth us trying to guard against.

Why not?  I.e. the cost seems to be 5 lines of boiler plate code that both document the intent and prevent any indirect mistake early.   And it is even already typed ... i.e. I don't see the cost :)

Reply:
It's mainly that I haven't seen explicit deletion for abstract base classes before. It seems like a lot of noise, and it could potentially hide the declaration of a default virtual destructor. So let's put all the "prevent slicing" code (everything but the destructor) into a single comment block:

```c++
    //!@{
    //! Prevent slicing by deleting move/copy operations
    AlongStepFactoryInterface() = default;
    AlongStepFactoryInterface(AlongStepFactoryInterface const&) = delete;
    AlongStepFactoryInterface& operator=(AlongStepFactoryInterface const&)
        = delete;
    AlongStepFactoryInterface(AlongStepFactoryInterface&&) = delete;
    AlongStepFactoryInterface& operator=(AlongStepFactoryInterface&&) = delete;
    //!@}
```

And maybe even move it to a `protected` block to keep it away from the main part of the class.

I mean, this is an unreadable wall of text that's going to be repeated through the code and should really be a single command/statement. Do we want to do something boost-y and have a helper class? e.g.
```c++
//! CRTP class for ABCs to prevent slicing
template<class T>
class Interface
{
protected:
    //!@{
    //! Prevent slicing by deleting move/copy operations
    Interface() = default;
    Interface(Interface const&) = delete;
    Interface& operator=(Interface const&) = delete;
    Interface(Interface&&) = delete;
    Interface& operator=(Interface&&) = delete;
    //!@}
};
```
and have all the `Interface` classes inherit from that (possibly privately?)

Reply:
Another alternative is to use a 'clearly' named macro to inject the code (It is not clear whether doing the deletion directly in place (via a macro) or indirectly via  base class is the clearer alternative.

Reply:
Is CRTP important here?

Reply:
Prevents casting one interface to another.